contracts/clients/CkbProof.sol

```diff
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import "../utils/Molecule.sol";
+import "solidity-rlp/contracts/RLPReader.sol";
+
+using Molecule for bytes;
+using RLPReader for bytes;
+using RLPReader for RLPReader.RLPItem;
+
+struct Proof {
+    uint32[] indices;
+    bytes32[] lemmas;
+    bytes32[] leaves;
+}
+
+struct VerifyProofPayload {
+    uint8 verifyType;
+    bytes32 transactionsRoot;
+    bytes32 witnessesRoot;
+    bytes32 rawTransactionsRoot;
+    Proof proof;
+}
+
+struct AxonObjectProof {
+    bytes ckbTransaction;
+    bytes32 blockHash;
+    VerifyProofPayload proofPayload;
+}
+
+struct CKBHeader {
+    uint32 version;
+    uint32 compactTarget;
+    uint64 timestamp;
+    uint64 number;
+    uint64 epoch;
+    bytes32 parentHash;
+    bytes32 transactionsRoot;
+    bytes32 proposalsHash;
+    bytes32 extraHash;
+    bytes32 dao;
+    uint128 nonce;
+    bytes extension;
+    bytes32 blockHash;
+}
+
+// Define the MsgType enum
+enum MsgType {
+    MsgClientCreate
+}
+
+// Define the CommitmentKV struct
+struct CommitmentKV {
+    uint256 key;
+    uint256 value;
+}
+
+// Define the Envelope struct
+struct Envelope {
+    MsgType msg_type;
+    CommitmentKV[] commitments;
+    bytes content;
+}
+
+contract CkbLightClient {
+    event GetHeaderEvent(CKBHeader);
+    event NotGetHeaderEvent();
+
+    function getHeader(
+        bytes32 blockHash
+    ) public virtual returns (CKBHeader memory) {
+        // axon_precompile_address(0x02)
+        address get_header_addr = address(0x0102);
+        (bool isSuccess, bytes memory res) = get_header_addr.staticcall(
+            abi.encode(blockHash)
+        );
+
+        CKBHeader memory header;
+        if (isSuccess) {
+            header = abi.decode(res, (CKBHeader));
+            emit GetHeaderEvent(header);
+        } else {
+            emit NotGetHeaderEvent();
+        }
+        return header;
+    }
+}
+
+// Define ckb blake2b
+contract Blake2b {
+    function blake2b(bytes memory data) public view returns (bytes32) {
+        // axon_precompile_address(0x06)
+        address blake2b_addr = address(0x0106);
+        (bool isSuccess, bytes memory res) = blake2b_addr.staticcall(
+            abi.encode(data)
+        );
+
+        bytes32 hash;
+        if (isSuccess) {
+            hash = abi.decode(res, (bytes32));
+        }
+        return hash;
+    }
+}
+
+contract CkbMbt {
+    function verify(
+        VerifyProofPayload memory payload
+    ) public view returns (bool) {
+        // axon_precompile_address(0x07)
+        address ckb_mbt_addr = address(0x0107);
+        (bool isSuccess, bytes memory res) = ckb_mbt_addr.staticcall(
+            abi.encode(payload)
```

Comment:
I had to use `abi.encode(payload.verifyType, payload.transactionsRoot, payload.witnessesRoot, payload.rawTransactionsRoot, payload.proof)` or there'll be some extra data in the front.

Reply:
If you use `abi.encode(payload)`, it's actually encoding the tuple with one element `(VerifyProofPayload)` so there will be an extra `head` because `VerifyProofPayload` is dynamic.

https://docs.soliditylang.org/en/develop/abi-spec.html#formal-specification-of-the-encoding

Reply:
Axon has agreed to change the input to `(VerifyProofPayload)`. We can change this back to `abi.encode(proof)` when it's done.